lib/lf/assembly/assembler.h

```diff
@@ -0,0 +1,219 @@
+/***************************************************************************
+ * LehrFEM++ - A simple C++ finite element libray for teaching
+ * Developed from 2018 at the Seminar of Applied Mathematics of ETH Zurich,
+ * lead developers Dr. R. Casagrande and Prof. R. Hiptmair
+ ***************************************************************************/
+
+/**
+ * @file
+ * @brief Functions for global finite element assembly of matrices and
+ *        right hand side vectors
+ * @author Ralf Hiptmair
+ * @date August 2018
+ * @copyright MIT License
+ */
+
+#ifndef _LF_ASSEMBLE_H
+#define _LF_ASSEMBLE_H
+
+#include <Eigen/Sparse>
+#include "dofhandler.h"
+#include "coomatrix.h"
+
+namespace lf::assemble {
+/**
+ * @brief Assembly function for standard assembly of finite element matrices
+ *
+ * @tparam CODIM co-dimension of mesh entities which should be traversed
+ *               in the course of assembly
+ * @tparam TMPMATRIX a type fitting the concept of COOMatrix
+ * @tparam ASSEMBLER a type providing the computation of element matrices
+ * @param dof_handler_trial a dof handler object for column space @see
+ * DofHandler
+ * @param dof_handler_test a dof handler object for row space @see DofHandler
+ * @param assembler assembler object for passing all kinds of data
+ * @param matrix matrix object to which the assembled matrix will be added.
+ *               The matrix object is not set to zero in the beginning!
+ *
+ * This method performs cell-oriented assembly controlled by a local-to-global
+ * index map ("dof handler").
+ *
+ * #### type requirements of template arguments
+ *
+ * - TMPMATRIX is a rudimentary matrix type and must
+ * + provide a constructor taking two matrix dimension arguments
+ * + have a method `AddtoEntry(i,j,value_to_add)` for adding to a matrix entry
+ * + be copyable and assignable
+ * A model type is COOMatrix.
+ * - ASSEMBLER is a type capable of local assembly of element matrices. It must
+ * + have an `Eval()` method returning the element matrix for a cell
+ * + supply an `isActive()` method for selecting cells to be taken into account
+ * in assembly
+ * + provide a matrix type `ElemMat` for objects containing the element
+ * matrices.
+ */
+template <int CODIM, typename TMPMATRIX, class ASSEMBLER>
+void AssembleMatrixLocally(const DofHandler &dof_handler_trial,
```

Comment:
Minor Remark: In Principle the `CODIM` parameter doesn't have to be a template parameter, it could also be a regular function parameter...

Reply:
You are right, CODIM should not be a template argument and can be passed as a regular argument. Can you please change and insert `dim_t codim` as first argument. Thanks.